hazelcast/src/main/java/com/hazelcast/map/impl/proxy/NearCachedMapProxyImpl.java

```diff
@@ -372,39 +372,49 @@ protected boolean containsKeyInternal(Object key) {
 
     @Override
     protected void getAllInternal(Set<K> keys, List<Data> dataKeys, List<Object> resultingKeyValuePairs) {
-        Map<Object, Data> keyMap = null;
         if (serializeKeys) {
-            keyMap = toDataKeys(keys, dataKeys);
+            toDataKeysWithReservations(keys, dataKeys, null, null);
         }
-        Collection ncKeys = serializeKeys ? dataKeys : new ArrayList<K>(keys);
+        Collection ncKeys = serializeKeys ? dataKeys : new LinkedList<K>(keys);
 
         populateResultFromNearCache(ncKeys, resultingKeyValuePairs);
         if (ncKeys.isEmpty()) {
             return;
         }
 
+        Map<Object, Long> reservations = createHashMap(ncKeys.size());
+        Map<Data, Object> reverseKeyMap = null;
         if (!serializeKeys) {
-            keyMap = toDataKeys(ncKeys, dataKeys);
+            reverseKeyMap = createHashMap(ncKeys.size());
+            toDataKeysWithReservations(ncKeys, dataKeys, reservations, reverseKeyMap);
+        } else {
+            createNearCacheReservations(ncKeys, reservations);
         }
 
-        Map<Object, Long> reservations = getNearCacheReservations(ncKeys, keyMap);
         try {
             int currentSize = resultingKeyValuePairs.size();
             super.getAllInternal(keys, dataKeys, resultingKeyValuePairs);
-            populateResultFromRemote(currentSize, resultingKeyValuePairs, reservations);
+            populateResultFromRemote(currentSize, resultingKeyValuePairs, reservations, reverseKeyMap);
         } finally {
             releaseReservedKeys(reservations);
         }
     }
 
-    private Map<Object, Data> toDataKeys(Collection<?> keys, Collection<Data> dataKeys) {
-        Map<Object, Data> keyMap = createHashMap(keys.size());
+    private void toDataKeysWithReservations(Collection<?> keys, Collection<Data> dataKeys, Map<Object, Long> reservations,
+                                            Map<Data, Object> reverseKeyMap) {
         for (Object key : keys) {
             Data keyData = toDataWithStrategy(key);
-            keyMap.put(key, keyData);
             dataKeys.add(keyData);
+            if (reservations != null) {
+                long reservationId = tryReserveForUpdate(key, keyData);
+                if (reservationId != NOT_RESERVED) {
+                    reservations.put(key, reservationId);
+                }
+            }
+            if (reverseKeyMap != null) {
+                reverseKeyMap.put(keyData, key);
+            }
         }
-        return keyMap;
     }
 
     private void populateResultFromNearCache(Collection keys, List<Object> resultingKeyValuePairs) {
```

Comment:
Just another curiosity - how come we call `toObject` when adding the results here into `resultingKeyValuePairs`? The rest of the code is equivalent to the cache proxy.
And `populateResultFromRemote` is expecting elements to be `Data`.
Maybe I'm just missing something here.